lib/cosmos/dart/lib/dart_common.rb

```diff
@@ -0,0 +1,524 @@
+      table_index += 1
+    end
+
+    # Mark packet config ready
+    packet_config.max_table_index = table_index - 1
+    packet_config.ready = true
+    packet_config.save
+  end
+
+  # Attempts to load a name system configuration. If the system configuration
+  # can't be loaded locally, it is requested from the server and copied
+  # locally before proceeding.
+  #
+  # @param system_config_name [String] System configuration name (MD5) to load
+  def switch_and_get_system_config(system_config_name)
+    # Switch to this new system configuration
+    current_config, error = Cosmos::System.load_configuration(system_config_name)
+
+    if current_config != system_config_name
+      Cosmos::Logger.warn("Failed to load system_config: #{system_config_name}")
+      Cosmos::Logger.warn("  Current config: #{current_config}")
+      Cosmos::Logger.warn("  Error: #{error.formatted}") if error
+      Cosmos::Logger.warn("  Will attempt to retrieve...")
+      filename, data = get_saved_config(system_config_name)
+      raise "No saved config" unless filename and data and data.length > 0
+      configuration = File.join(Cosmos::System.paths['SAVED_CONFIG'], filename)
+      unless File.exist?(configuration) and File.size(configuration) > 0
+        File.open(configuration, 'wb') {|file| file.write(data)}
+        File.chmod(0444, configuration) # Mark readonly
+      end
+      Cosmos::Logger.info("Configuration retrieved: #{configuration}")
+      current_config, error = Cosmos::System.load_configuration(system_config_name)
+      raise "Could not load config" if current_config != system_config_name
+    end
+  end
+
+  # Iterate through all the defined commands and telemetry and create database
+  # entries for all targets (Target table) and packets (Packet table).
+  def sync_targets_and_packets
+    sync_targets_packets(Cosmos::System.telemetry.all, is_tlm: true)
+    sync_targets_packets(Cosmos::System.commands.all, is_tlm: false)
+  end
+
+  # Build the internal lookup tables to convert names to database ids
+  def build_lookups
+    # Get full target name and packet name lists from database
+    targets = Target.all
+    @target_name_to_id = {}
+    @target_id_tlm_packet_name_to_id = {}
+    @target_id_cmd_packet_name_to_id = {}
+    @packet_id_item_name_to_id = {}
+    targets.each do |target|
+      @target_name_to_id[target.name] = target.id
+      @target_id_tlm_packet_name_to_id[target.id] = {}
+      @target_id_cmd_packet_name_to_id[target.id] = {}
+      packets = Packet.where("target_id = #{target.id} and is_tlm = true").all
+      packets.each do |packet|
+        @target_id_tlm_packet_name_to_id[target.id][packet.name] = packet.id
+        @packet_id_item_name_to_id[packet.id] = {}
+        items = Item.where("packet_id = #{packet.id}").all
+        items.each do |item|
+          @packet_id_item_name_to_id[packet.id][item.name] = item.id
+        end
```

Comment:
build_lookups is always called after sync_targets_and_packets so I would recommend moving it to the end of sync_targets_and_packets so client code doesn't have to remember to do it. Also, this block of code starting with Item.where basically doesn't do anything because no Items have been created. Only when you call sync_item does an Item get created. sync_item is only called by lookup_item_id, which is only called by setup_item_to_decom_table_mapping, which is only called by setup_packet_config. setup_packet_config is conditionally called by Dart cleanup and DartWorker initialize. Is there a reason sync_targets_and_packets doesn't just sync all the items? You're going to have to do it eventually during the startup process right?

Reply:
Ok so syncing the Target and Packets is an optimization (plus creating a PacketLogEntry needs them) and Items is only for decommutation.